### PR TITLE
Pwm improvements

### PR DESCRIPTION
**Pull request by hmvp**
I was not entirely satisfied with the way PWM worked so I started to modify it a bit. That turned into a full rewrite. However I am not satisfied yet.

The upside is that this models the actual peripheral better where some settings are shared between channels and multiple pins can be used simultaneous by the same channel.

The downside is that you might want to "start" with the gpio pin and "just" find the right pwm slice that belongs to that. I found no way to do that except for an macro that needs to be used in the same scope as the `Slices` instance..
Another issue is that once you borrow a channel (like in the example just before the loop) the slice is no longer available to change settings. We could probably change that by making all methods take `&self` instead of `&mut self` but I am not sure whats best here...

Any insights are welcome

### CODE REVIEW

**Comment by hmvp** [comment]
@anall I took a lot of inspiration from the gpio system. Maybe you have some comments?

**Comment by anall** [comment]
> @anall I took a lot of inspiration from the gpio system. Maybe you have some comments?

I'll try and take a look this weekend.

File: rp2040-hal/src/pwm/mod.rs
Comment (henkkuli) [inline rp2040-hal/src/pwm/mod.rs:3]: This documentation along with the examples needs to be updated.

File: rp2040-hal/src/pwm/mod.rs
Comment (henkkuli) [inline rp2040-hal/src/pwm/mod.rs:156]: The documentation could contain the numeric ID.
```suggestion
        #[doc = std::concat!("Slice ID representing slice ", $NUM)]
```

File: rp2040-hal/src/pwm/mod.rs
Comment (henkkuli) [inline rp2040-hal/src/pwm/mod.rs:173]:
```suggestion
/// access the corresponding registers.
```

File: rp2040-hal/src/pwm/mod.rs
Comment (henkkuli) [inline rp2040-hal/src/pwm/mod.rs:311]:
```suggestion
/// Marker trait for valid PWM input pins.
```

File: rp2040-hal/src/pwm/mod.rs
Comment (henkkuli) [inline rp2040-hal/src/pwm/mod.rs:313]:
```suggestion
/// Marker trait for valid PWM output pins.
```

File: rp2040-hal/src/pwm/mod.rs
Comment (henkkuli) [inline rp2040-hal/src/pwm/mod.rs:317]: Is this trait used anywhere? How about `PwmInputChannel`?

File: rp2040-hal/src/pwm/mod.rs
Comment (henkkuli) [inline rp2040-hal/src/pwm/mod.rs:415]: Why using `#[doc = ""]` and not `///`? Also
```suggestion
    #[doc = "Enable phase correct mode"]
```

File: rp2040-hal/src/pwm/mod.rs
Comment (henkkuli) [inline rp2040-hal/src/pwm/mod.rs:517]: This should probably return `PwmPinToken` because otherwise there is no way to recover the pin back from PWM.

File: rp2040-hal/src/pwm/dynslice.rs
Comment (henkkuli) [inline rp2040-hal/src/pwm/dynslice.rs:43]:
```suggestion
/// access the corresponding registers.
```

File: rp2040-hal/src/pwm/dynslice.rs
Comment (henkkuli) [inline rp2040-hal/src/pwm/dynslice.rs:150]:
```suggestion
    /// Enable slice
```

**Review by henkkuli — COMMENTED** [review]
Even though I'm not a maintainer of this crate, I decided to leave a few comments on this PR as I need it. In general I like what this PR is doing and the API matches my expectations.

As a general note, all public methods should have documentation. Now some methods have empty doc string.

File: rp2040-hal/src/pwm/reg.rs
Comment (henkkuli) [inline rp2040-hal/src/pwm/reg.rs:86]: This overwrites the value written to channel b. Use
```suggestion
        ch.cc.modify(|_, w| unsafe { w.a().bits(value) });
```
instead

**Comment by hmvp** [comment]
> Even though I'm not a maintainer of this crate, I decided to leave a few comments on this PR as I need it. In general I like what this PR is doing and the API matches my expectations.
> 
That's cool, thanks for the comments. I wasn't nearly finished with this so there are some obvious omissions as you pointed out. 
I put it out here in this because I have some doubts about the direction (See first comment) and didn't want to finish something that might be the wrong direction al together.. 

@9names @anall Did you have a chance to look at the direction I took this?

**Comment by 9names** [comment]
I had a look, and I like the API you're aiming for better than the previous one.
The original pin-level PWM abstraction really doesn't match the hardware well

I'd would like to know the reason why you want type-erased versions of PWM slices though.
I understand the cost/benefit in doing this for GPIO/pins (though I do not like the cost), but I'm not convinced that there is a *need* for this for PWM slices.
Happy to have this if it solves a problem, I just don't know what that problem is right now.
Also: I'm not nearly confident enough in type-state programming to provide any criticism or help here.

As far as interaction with the PWM peripheral goes:
I'm reading it correctly you "disable" a channel by setting it's duty cycle to 0. A common way to disable a single channel of a PWM pair when there is no channel disable bit is to disconnect the pin from the peripheral, so that the channel output is no longer driving the pin.
This relieves you of the need to store the duty cycle anywhere except the register, and is immediate.

**Comment by hmvp** [comment]
> I had a look, and I like the API you're aiming for better than the previous one.
> The original pin-level PWM abstraction really doesn't match the hardware well
> 
Yeah
> I'd would like to know the reason why you want type-erased versions of PWM slices though.
> I understand the cost/benefit in doing this for GPIO/pins (though I do not like the cost), but I'm not convinced that there is a _need_ for this for PWM slices.
> Happy to have this if it solves a problem, I just don't know what that problem is right now.
> Also: I'm not nearly confident enough in type-state programming to provide any criticism or help here.
> 
To be honest I mostly copied gpio. But I'm still struggling with two issues: 
- One is how to get a pwm slice/channel based on a gpio pin. It seems that would make it more user friendly.
- The other one is the enable multiple slices at the same time. I have not found a way to return multiple slices in some way..
It feels a bit like the Dyn stuff might help with that. Its just that am still looking to see if I can check that at compile time (using types)

> As far as interaction with the PWM peripheral goes:
> I'm reading it correctly you "disable" a channel by setting it's duty cycle to 0. A common way to disable a single channel of a PWM pair when there is no channel disable bit is to disconnect the pin from the peripheral, so that the channel output is no longer driving the pin.
> This relieves you of the need to store the duty cycle anywhere except the register, and is immediate.
That is a better solution, but you need to know which pin is used (and what to set it to). all channels can output to two pins. Even both at the same time... Maybe I should implement the PwmPin trait on the PwmPinToken... But that makes it harder to implement working on both pins...

File: rp2040-hal/src/pwm/dynslice.rs
Comment (hmvp) [reply rp2040-hal/src/pwm/dynslice.rs:159]: haha yes, I am not even ready to actually test the stuff yet..

**Comment by hmvp** [comment]
I pushed a rebased and heavily changed version. I had more time to test this and am reasonably happy with it.. The two commented functions are still something I feel might need to added. But I cannot get it to work with Rusts constraints...

**Comment by anall** [comment]
I'll try to give this a look later this week. My day job and other commitments have been very demanding recently.

> I'd would like to know the reason why you want type-erased versions of PWM slices though.
> I understand the cost/benefit in doing this for GPIO/pins (though I do not like the cost)

I believe in a lot of cases the optimizer should get rid of the runtime cost (at least for GPIO) if you are not using the type-erased variants. There will obviously be a runtime cost if the type-erased version is used.

But I agree it is probably overkill if it is not needed (I have yet to give this PR a look)

>One is how to get a pwm slice/channel based on a gpio pin. It seems that would make it more user friendly.

I'll have to give this a deeper look, but I feel like there is something clever you can do.

File: rp2040-hal/src/pwm/mod.rs
Comment (9names) [reply rp2040-hal/src/pwm/mod.rs:3]: Where is the split method? What is a `_____` struct?
I think this first bit of the docs needs little more attention.

File: rp2040-hal/src/pwm/mod.rs
Comment (9names) [inline rp2040-hal/src/pwm/mod.rs:58]: I think it would be good to show here how to bring the pwm slice into one of the other modes as well, since it's not obvious.
```rust
/// Set this channel 
let pwm: &mut hal::pwm::Slice<hal::pwm::Pwm4, hal::pwm::InputHighRunning> =
        &mut pwm_slices.pwm4.into_mode();
```

**Review by 9names — CHANGES_REQUESTED** [review]
I finally found enough time to test and review this code - sorry about the delay!
The implementation seems quite robust and easy to use, I really like that you can set multiple input/output pins easily.
Happy to merge this once you've improved the docs where requested.

**Review by 9names — APPROVED** [review]
LGTM

**Pull request merged**
